lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                if (params && params.versionId) {
+                    // eslint-disable-next-line
+                    key = formatVersionKey(objName, params.versionId, vFormat);
+                } else {
+                    // eslint-disable-next-line
+                    key = formatMasterKey(objName, vFormat);
+                }
+                c.findOne({
+                    _id: key,
+                }, {}, (err, doc) => {
                     if (err) {
-                        log.error('getLatestVersion: getting latest version',
+                        log.error('findOne: error getting object',
                             { error: err.message });
-                        return cb(err);
+                        return next(errors.InternalError);
                     }
-                    return cb(null, value);
+                    if (!doc) {
+                        return next(errors.NoSuchKey);
+                    }
+                    if (doc.value.isPHD) {
+                        this.getLatestVersion(c, objName, vFormat, log, (err, value) => {
+                            if (err) {
+                                log.error('getLatestVersion: getting latest version',
+                                    { error: err.message });
+                                return next(err);
+                            }
+                            return next(null, value);
+                        });
+                        return undefined;
+                    }
+                    MongoUtils.unserialize(doc.value);
+                    return next(null, doc.value);
                 });
-                return undefined;
-            }
-            MongoUtils.unserialize(doc.value);
-            return cb(null, doc.value);
-        });
+            },
+        ], cb);
     }
 
-    /*
-     * This function return the latest version
+    /**
+     * This function return the latest version of an object
+     * by getting all keys related to an object, ordering them
+     * and returning the latest one
+     * @param {Object} c collection
+     * @param {String} objName object name
+     * @param {String} vFormat bucket version format
+     * @param {Object} log logger
+     * @param {Function} cb callback
+     * @return {undefined}
      */
-    getLatestVersion(c, objName, log, cb) {
+    getLatestVersion(c, objName, vFormat, log, cb) {
+        // generating the range delimiter keys
+        const masterKey = formatMasterKey(objName, vFormat);
+        // version id is added at the end of the key so giving it an empty
+        // string gives us the last key in the range
+        const versionKey = formatVersionKey(objName, VID_NONE, vFormat);
+        const lastVersionKey = inc(versionKey);
+        const filter = {};
+        if (vFormat === BUCKET_VERSIONS.v0) {
+            filter.$gt = masterKey;
+            filter.$lt = lastVersionKey;
+        } else {
+            filter.$gt = versionKey;
```

Comment:
It may not be enough to retrieve only the version key in v1, due to the fact that a bucket may be nonversioned for some time and only create master keys without associated version keys. I think we need to check both prefixes for that reason in v1.

Reply:
`getLatestVersion` only returns the versions and never the master, this is the case in previous implementation as well. You can see that it gets keys greater than the master key and not greater or equal.
If i changed the function to also check the master prefix it would mess up the other functions calling it. An example would be the repair function that gets the latest version to decide if it should repair or delete the master.
Having the function in it's current state doesn't create a problem when retrieving objects that don't have a version associated to them as we get them by their master key in `getObject`